Sources/MapboxDirections/RouteRoadClassesViolations.swift

```diff
@@ -0,0 +1,85 @@
+
+import Foundation
+
+/**
+ Description of `RoadClasses` that were meant to be avoided during routing but still got used in a `Route`.
+ 
+ If something in `RouteOptions.roadClassesToAvoid` prevents the origin or destination from connecting to the main road network, the Directions API will temporarily ignore the road class constraints in order to reach the main road network.
+ */
+public struct RouteRoadClassesViolations {
+    /**
+     Related `Route` object
+     */
+    public let route: Route
+    
+    /**
+     List of all existing violations, down to `RouteStep` granularity.
+     */
+    public let violations: [RoadClassExclusionViolation]
+    
+    /**
+     Filters `violations` to search for specific leg and step.
+     
+     - parameter legIndex: Index of a leg inside related `Route`to search in.
+     - parameter stepIndex: Index of a step inside given `Route`'s leg.
+     - returns: Array of `RoadClassExclusionViolation` under given indicies.
+     
+     Passing `nil` as `stepIndex` will result in searching for all steps.
+     */
+    public func violations(at legIndex: Int, stepIndex: Int? = nil) -> [RoadClassExclusionViolation] {
+        return fetchViolations(at: legIndex, stepIndex: stepIndex, intersectionIndex: nil)
+    }
+    
+    /**
+     Filters `violations` to search for specific leg, step and intersection.
+     
+     - parameter legIndex: Index of a leg inside related `Route`to search in.
+     - parameter stepIndex: Index of a step inside given `Route`'s leg.
+     - parameter intersectionIndex: Index of an intersection inside given `Route`'s leg and step.
+     - returns: Array of `RoadClassExclusionViolation` under given indicies.
+     
+     Passing `nil` as `intersectionIndex` will result in searching for all intersections of given step.
+     */
+    public func violations(at legIndex: Int, stepIndex: Int, intersectionIndex: Int?) -> [RoadClassExclusionViolation] {
+        return fetchViolations(at: legIndex, stepIndex: stepIndex, intersectionIndex: intersectionIndex)
+    }
+    
+    private func fetchViolations(at legIndex: Int, stepIndex: Int? = nil, intersectionIndex: Int? = nil) -> [RoadClassExclusionViolation] {
+        assert(!(stepIndex == nil && intersectionIndex != nil), "It is forbidden to select `intersectionIndex` without specifying `stepIndex`.")
+        
+        return violations.filter { violation in
+            let validStep = (stepIndex ?? violation.stepIndex) == violation.stepIndex
+            let validIntersection = (intersectionIndex ?? violation.intersectionIndex) == violation.intersectionIndex
+            
+            return violation.legIndex == legIndex &&
+                validStep &&
+                validIntersection
+        }
```

Comment:
This logic works, but it would be simpler if we break it down into multiple lazy filters: first lazily filter by leg, then if a step index is specified, lazily filter by step, then if an intersection index, filter by intersection.